core/bin/snapshots_creator/src/main.rs

```diff
@@ -0,0 +1,306 @@
+        .put(l1_batch_number, &factory_deps)
+        .await
+        .context("Error storing factory deps in blob store")?;
+    let output_filepath_prefix = blob_store.get_storage_prefix::<SnapshotFactoryDependencies>();
+    let output_filepath = format!("{output_filepath_prefix}/{filename}");
+    let elapsed_ms = latency.observe().as_millis();
+    tracing::info!(
+        "Finished factory dependencies, step took {elapsed_ms}ms , output stored in {}",
+        output_filepath
+    );
+    Ok(output_filepath)
+}
+
+async fn run(
+    blob_store: Box<dyn ObjectStore>,
+    replica_pool: ConnectionPool,
+    master_pool: ConnectionPool,
+) -> anyhow::Result<()> {
+    let latency = METRICS.snapshot_generation_duration.start();
+
+    let config = SnapshotsCreatorConfig::from_env().context("SnapshotsCreatorConfig::from_env")?;
+
+    let mut conn = replica_pool
+        .access_storage_tagged("snapshots_creator")
+        .await?;
+
+    // we subtract 1 so that after restore, EN node has at least one l1 batch to fetch
+    let l1_batch_number = conn.blocks_dal().get_sealed_l1_batch_number().await? - 1;
+
+    let mut master_conn = master_pool
+        .access_storage_tagged("snapshots_creator")
+        .await?;
+    if master_conn
+        .snapshots_dal()
+        .get_snapshot_metadata(l1_batch_number)
+        .await?
+        .is_some()
+    {
+        tracing::info!("Snapshot for L1 batch number {l1_batch_number} already exists, exiting",);
+        return Ok(());
+    }
+    drop(master_conn);
+
+    // snapshots always
+    let last_miniblock_number_in_batch = conn
+        .blocks_dal()
+        .get_miniblock_range_of_l1_batch(l1_batch_number)
+        .await?
+        .context("Error fetching last miniblock number")?
+        .1;
+    let distinct_storage_logs_keys_count = conn
+        .snapshots_creator_dal()
+        .get_distinct_storage_logs_keys_count(l1_batch_number)
+        .await?;
+
+    drop(conn);
+
+    let chunk_size = config.storage_logs_chunk_size;
+    // we force at least 10 chunks to avoid situations where only one chunk is created in tests
+    let chunks_count = max(10, ceil_div(distinct_storage_logs_keys_count, chunk_size));
+
+    METRICS.storage_logs_chunks_count.set(chunks_count);
+
+    tracing::info!(
+        "Creating snapshot for storage logs up to miniblock {last_miniblock_number_in_batch}, l1_batch {}",
+        l1_batch_number.0
+    );
+    tracing::info!("Starting to generate {chunks_count} chunks of expected size {chunk_size}");
+
+    let factory_deps_output_file = process_factory_deps(
+        &*blob_store,
+        &replica_pool,
+        last_miniblock_number_in_batch,
+        l1_batch_number,
+    )
+    .await?;
+
+    let mut storage_logs_output_files = vec![];
+
+    METRICS
+        .storage_logs_chunks_left_to_process
+        .set(chunks_count);
+    let mut tasks =
+        FuturesUnordered::<Pin<Box<dyn Future<Output = anyhow::Result<String>>>>>::new();
```

Comment:
hmm interesting - I don't have enough rust experience to reason about what's more idiomatic, but from a quick glance, the current solution seems way too imperative, I would think it's possible to use a higher-level abstraction. Again, from a quick glance `for_each_concurrent` seems perfect, but a semaphore solution is also simpler than this

For me it's a little bit like using a while loop and manipulating some counter directly instead of a higher-order function like `map`.

> which I can monitor progress and from which I can pull finished tasks, that's why I also don't want to use join_all as I want to know exactly when tasks finish.

Not sure how exactly this applies here - can you please elaborate?

Reply:
I meant that I do not see any nice functional version of this code where we can track progress, but I also did not take `map` with side effects as a consideration. I usually try to avoid using it with functions that are not 'pure'. Also, the part that uses the counter is not thread-safe, so we would need a second semaphore (or rather a mutex). TLDR; I think the devil is in the details and that imperative version is the lesser evil, but would very like to be proven that this is not the case ;). @slowli can you show me a snippet or even pseudocode how you imagine the code using semaphore(s) would be structured?

Reply:
The core code is as follows:

```rust
let semaphore = Semaphore::new(config.concurrent_queries_count as usize);
let tasks = (0..chunks_count).map(|chunk_id| {
    process_storage_logs_single_chunk(
        &semaphore,
        &*blob_store,
        &replica_pool,
        last_miniblock_number_in_batch,
        l1_batch_number,
        chunk_id,
        chunks_count,
    )
});
let storage_logs_output_files = futures::future::try_join_all(tasks).await?;
```

...where `process_storage_logs_single_chunk` is modified to acquire a `semaphore` permit:

```rust
async fn process_storage_logs_single_chunk(
    semaphore: &Semaphore,
    // other args...
) -> anyhow::Result<String> {
    let _permit = semaphore.acquire().await?;
    // ...
}
```

> Also, the part that uses the counter is not thread-safe

Why would we need a counter in the first place? Theoretically, we could use an atomic, but if it's only required for metrics and logging, then you should use [`Gauge::dec_by()`](https://matter-labs.github.io/vise/vise/struct.Gauge.html#method.dec_by) (which uses an atomic internally).

A variant with `Stream`s:

```rust
let tasks = futures::stream::iter(0..chunks_count).map(|chunk_id| {
    process_storage_logs_single_chunk(
        &*blob_store,
        &replica_pool,
        last_miniblock_number_in_batch,
        l1_batch_number,
        chunk_id,
        chunks_count,
    )
});
let storage_logs_output_files: Vec<_> = tasks
    .buffer_unordered(config.concurrent_queries_count as usize)
    .try_collect()
    .await?;
```

Reply:
corrected